mm/tempesta_mm.c

```diff
@@ -196,30 +101,42 @@ void __init
 tempesta_reserve_pages(void)
 {
 	int nid;
-	struct page *p;
+	void *addr;
+	void *pos, *end;
 
 	for_each_online_node(nid) {
-		p = tempesta_alloc_contmem(nid);
-		if (!p)
+		addr = memblock_alloc_try_nid_raw(dbmem, HPAGE_SIZE,
+						  MEMBLOCK_LOW_LIMIT,
+						  MEMBLOCK_ALLOC_ANYWHERE,
+						  nid);
+		if (!addr) {
+			pr_err("Tempesta: can't reserve %lu memory at node %d"
+			       "\n", __dbsize_mb(dbmem), nid);
 			goto err;
+		}
 
-		map[nid].addr = (unsigned long)page_address(p);
-		map[nid].pages = PGNUM4K;
+		end = addr + dbmem;
+		for (pos = addr; pos < end; pos += HPAGE_SIZE)
+			prep_huge_compound_page(virt_to_page(pos),
+						HUGETLB_PAGE_ORDER);
 
+		map[nid].addr = (unsigned long)addr;
+		map[nid].pages = dbmem / PAGE_SIZE;
 		pr_info("Tempesta: allocated huge pages space %pK %luMB at node"
-			" %d\n", page_address(p),
-			PGNUM4K * PAGE_SIZE / (1024 * 1024), nid);
+			" %d\n", addr, __dbsize_mb(dbmem), nid);
 	}
 
 	return;
+
 err:
 	for_each_online_node(nid) {
-		struct page *pend;
+		phys_addr_t phys_addr;
+
 		if (!map[nid].addr)
 			continue;
-		for (p = virt_to_page(map[nid].addr), pend = p + PGNUM4K;
-		     p < pend; p += 1 << HUGETLB_PAGE_ORDER)
-			tempesta_free_hpage(p);
+
+		phys_addr = virt_to_phys((void *)map[nid].addr);
+		memblock_free(phys_addr, map[nid].pages * PAGE_SIZE);
```

Comment:
[destroy_compound_gigantic_page](https://elixir.bootlin.com/linux/v5.10.35/C/ident/destroy_compound_gigantic_page) It seems we should implement this function in our code or use prep_compound_gigantic_page/destroy_compound_gigantic_page directly (make them non static)
memblock_free is not enough